core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -0,0 +1,281 @@
+import org.apache.spark.internal.config._
+import org.apache.spark.scheduler._
+import org.apache.spark.storage.RDDBlockId
+import org.apache.spark.util.Clock
+
+/**
+ * A monitor for executor activity, used by ExecutorAllocationManager to detect idle executors.
+ */
+private[spark] class ExecutorMonitor(
+    conf: SparkConf,
+    client: ExecutorAllocationClient,
+    clock: Clock) extends SparkListener with Logging {
+
+  private val idleTimeoutMs = TimeUnit.SECONDS.toMillis(
+    conf.get(DYN_ALLOCATION_EXECUTOR_IDLE_TIMEOUT))
+  private val storageTimeoutMs = TimeUnit.SECONDS.toMillis(
+    conf.get(DYN_ALLOCATION_CACHED_EXECUTOR_IDLE_TIMEOUT))
+  private val fetchFromShuffleSvcEnabled = conf.get(SHUFFLE_SERVICE_ENABLED) &&
+    conf.get(SHUFFLE_SERVICE_FETCH_RDD_ENABLED)
+
+  private val executors = new ConcurrentHashMap[String, Tracker]()
+
+  // Keep track of the time the next executor is expected to time out. This is an optimization to
+  // avoid having to scan all executors on every EAM schedule interval.
+  //
+  // There's also some per-executor state used to decide when to update this value, so that
+  // recomputations can be triggered only when really necessary.
+  //
+  // Note that this isn't meant to, and cannot, always make the right decision about which executors
+  // are indeed timed out. For example, the EAM thread may detect a timed out executor while a new
+  // "task start" event has just been posted to the listener bus and hasn't yet been delivered to
+  // this listener. There are safeguards in other parts of the code that would prevent that executor
+  // from being removed.
+  private var nextTimeout = new AtomicLong(Long.MaxValue)
+
+  def reset(): Unit = {
+    executors.clear()
+    nextTimeout.set(Long.MaxValue)
+  }
+
+  /**
+   * Returns the list of executors that are currently considered to be timed out.
+   * Should only be called from the EAM thread.
+   */
+  def timedOutExecutors(): Seq[String] = {
+    val now = clock.getTimeMillis()
+    if (now >= nextTimeout.get()) {
+      // Temporarily set the next timeout at Long.MaxValue. This ensures that after
+      // scanning all executors below, we know when the next timeout for non-timed out
+      // executors is (whether that update came from the scan, or from a new event
+      // arriving in a different thread).
+      nextTimeout.set(Long.MaxValue)
+
+      var newNextTimeout = Long.MaxValue
+      val timedOutExecs = executors.asScala
+        .filter { case (_, exec) => !exec.pendingRemoval }
+        .filter { case (_, exec) =>
+          val deadline = exec.timeoutAt
+          if (deadline > now) {
+            newNextTimeout = math.min(newNextTimeout, deadline)
+            exec.timedOut = false
+            false
+          } else {
+            exec.timedOut = true
+            true
+          }
+        }
+        .keys
+        .toSeq
+
+      updateNextTimeout(newNextTimeout)
+      timedOutExecs
+    } else {
+      Nil
```

Comment:
I actually thought it was more clear with the cached value, but fine with this too

Reply:
I actually found some issue with the current code after I removed another piece of code that seemed to be doing nothing useful. That might also be related to the unit test failure in the last run. Taking a look...